src/debugpy/server/tracing/tracer.py

```diff
@@ -0,0 +1,417 @@
+        self.log.exception("{0}", *args, **kwargs)
+
+
+class Tracer:
+    """
+    Singleton that manages sys.monitoring callbacks for this process.
+    """
+
+    import inspect
+    import threading
+    from debugpy import server
+    from debugpy.server.tracing import Breakpoint, Step, Thread, StackFrame, _cvar
+    from pathlib import Path
+    from sys import monitoring
+
+    instance: "Tracer"
+
+    log: Log
+
+    _pause_ids = set()
+    """IDs of threads that are currently pausing or paused."""
+
+    _steps = {}
+    """Ongoing steps, keyed by thread ID."""
+
+    def __init__(self):
+        self.log = Log()
+
+    @property
+    def adapter(self):
+        return self.server.adapter()
+
+    def start(self):
+        """
+        Register sys.monitoring tracing callbacks.
+        """
+
+        self.log.info("Registering sys.monitoring tracing callbacks...")
+
+        self.monitoring.use_tool_id(self.monitoring.DEBUGGER_ID, "debugpy")
+        self.monitoring.set_events(
+            self.monitoring.DEBUGGER_ID,
+            (
+                self.monitoring.events.LINE
+                | self.monitoring.events.PY_START
+                | self.monitoring.events.PY_RETURN
+                | self.monitoring.events.PY_RESUME
+                | self.monitoring.events.PY_YIELD
+                | self.monitoring.events.PY_THROW
+                | self.monitoring.events.PY_UNWIND
+                | self.monitoring.events.RAISE
+                | self.monitoring.events.RERAISE
+                | self.monitoring.events.EXCEPTION_HANDLED
+            ),
+        )
+        trace_funcs = {
+            self.monitoring.events.LINE: self._trace_line,
+            self.monitoring.events.PY_START: self._trace_py_start,
+            self.monitoring.events.PY_RESUME: self._trace_py_resume,
+            self.monitoring.events.PY_RETURN: self._trace_py_return,
+            self.monitoring.events.PY_YIELD: self._trace_py_yield,
+            self.monitoring.events.PY_THROW: self._trace_py_throw,
+            self.monitoring.events.PY_UNWIND: self._trace_py_unwind,
+            self.monitoring.events.RAISE: self._trace_raise,
+            self.monitoring.events.RERAISE: self._trace_reraise,
+            self.monitoring.events.EXCEPTION_HANDLED: self._trace_exception_handled,
+        }
+        for event, func in trace_funcs.items():
+            self.monitoring.register_callback(self.monitoring.DEBUGGER_ID, event, func)
+
+        self.log.info("sys.monitoring tracing callbacks registered.")
+
+    def pause(self, thread_ids: Iterable[int] = None):
+        """
+        Pause the specified threads, or all threads if thread_ids is None.
+        """
+        if thread_ids is None:
+            # Pausing is async, so additional threads may be spawned even as we are
+            # trying to pause the ones we currently know about; iterate until all
+            # known threads are paused, and no new threads appear.
```

Comment:
Without this loop, the race will result in not all threads being paused occasionally, which breaks DAP invariants when "pause" request reports all threads to be paused (and VS in particular *requires* this), so it is crucial.

The loop should be robust in the face of any real-world multithreaded code. One case where in theory it can never terminate is if the code running is some kind of fork bomb, spawning threads which immediately spawn more threads etc; although even in this case, given that spawning a thread is significantly more expensive than a single iteration of the loop, I would expect it to converge eventually.

Reply:
I was thinking more of a thread starting as the loop gets exited, so not everything is paused on exit.

Reply:
Yes, that's a good point. Now that I think about it, it's possible for a thread to be initializing, but not yet at the point where `enumerate()` would see it. This needs an additional check in line tracing so that any new thread that is spawned while paused also gets paused as soon as it starts running Python code.

Reply:
So I tried to implement this and ran into two fundamental issues.

The first one is that DAP is simply underspecified for various corner cases that arise when pause/unpause is supported both for all threads and for specific threads - e.g. if all threads are paused, but then a single thread is resumed, and that thread spawns a new thread, should the new thread be paused or resumed? There are similar issues with evaluate etc.

The other problem is that VS (unlike VSCode) simply doesn't support individual thread requests and requires all threads to be paused/resumed, otherwise it gets out of sync. And while DAP has a way for the client to communicate that it supports single-thread-continue, there's no equivalent for single-thread-pause; thus we have to always assume that the client is VS (or something with the same limitation) and always pause all threads.

Between the two, I'm going to simplify this by tracking paused/unpaused state globally and not supporting it for individual threads. This means that in VSCode, if you right-click a single thread and ask to unpause it, it'll unpause all threads in practice. I believe this is acceptable as that is how pydevd-based debugpy behaves already. While there has been a user request to add per-thread unpause (#1475), I don't think we can do a good job there until DAP spec is extended & clarified sufficiently for us to implement; https://github.com/microsoft/debug-adapter-protocol/issues/355 is the issue tracking this on DAP side.